File: NC-150/08_Tries/62_add_and_search_word.py
"""
LeetCode 211: Design Add and Search Words Data Structure
https://leetcode.com/problems/add-and-search-word-data-structure-design/

Design a data structure that supports adding new words and finding if a 
string matches any previously added string.

Implement the WordDictionary class:
- WordDictionary() Initializes the object.
- void addWord(word) Adds word to the data structure.
- bool search(word) Returns true if there is any string in the data structure 
  that matches word or false otherwise. word may contain dots '.' where dots 
  can be matched with any letter.

Time Complexity: 
  - addWord: O(m) where m is word length
  - search: O(m) for words without '.', O(26^m) worst case with '.'
Space Complexity: O(n * m) where n is number of words
"""


class TrieNode:
    def __init__(self):
        self.children = {}  # char -> TrieNode
        self.is_end_of_word = False


class WordDictionary:
    def __init__(self):
        self.root = TrieNode()

    def addWord(self, word: str) -> None:
        node = self.root
        for char in word:
            if char not in node.children:
                node.children[char] = TrieNode()
            node = node.children[char]
        node.is_end_of_word = True

    def search(self, word: str) -> bool:
        def search_helper(index: int, node: TrieNode) -> bool:
            if index == len(word):
                return node.is_end_of_word
            
            char = word[index]
            
            if char == '.':
                # Wildcard: try all possible children
                for child in node.children.values():
                    if search_helper(index + 1, child):
                        return True
                return False
            else:
                # Regular character
                if char not in node.children:
                    return False
                return search_helper(index + 1, node.children[char])
        
        return search_helper(0, self.root)


# Alternative solution using iterative approach with stack
class WordDictionaryIterative:
    def __init__(self):
        self.root = TrieNode()

    def addWord(self, word: str) -> None:
        node = self.root
        for char in word:
            if char not in node.children:
                node.children[char] = TrieNode()
            node = node.children[char]
        node.is_end_of_word = True

    def search(self, word: str) -> bool:
        # Stack of (node, index) pairs
        stack = [(self.root, 0)]
        
        while stack:
            node, index = stack.pop()
            
            if index == len(word):
                if node.is_end_of_word:
                    return True
                continue
            
            char = word[index]
            
            if char == '.':
                # Add all children to stack
                for child in node.children.values():
                    stack.append((child, index + 1))
            else:
                if char in node.children:
                    stack.append((node.children[char], index + 1))
        
        return False


# Test
if __name__ == "__main__":
    word_dict = WordDictionary()
    
    word_dict.addWord("bad")
    word_dict.addWord("dad")
    word_dict.addWord("mad")
    
    print(f"search('pad'): {word_dict.search('pad')}")  # False
    print(f"search('bad'): {word_dict.search('bad')}")  # True
    print(f"search('.ad'): {word_dict.search('.ad')}")  # True
    print(f"search('b..'): {word_dict.search('b..')}")  # True
    print(f"search('b.d'): {word_dict.search('b.d')}")  # True
    print(f"search('...'): {word_dict.search('...')}")  # True
    print(f"search('..'): {word_dict.search('..')}")    # False
    
    # Test iterative version
    print("\nIterative version:")
    word_dict_iter = WordDictionaryIterative()
    word_dict_iter.addWord("bad")
    word_dict_iter.addWord("dad")
    print(f"search('.ad'): {word_dict_iter.search('.ad')}")  # True

File: NC-150/08_Tries/62_add_and_search_word.cpp
/*
 * LeetCode 211: Design Add and Search Words Data Structure
 * https://leetcode.com/problems/add-and-search-word-data-structure-design/
 * 
 * Design a data structure that supports adding new words and finding if a 
 * string matches any previously added string.
 * 
 * Implement the WordDictionary class:
 * - WordDictionary() Initializes the object.
 * - void addWord(word) Adds word to the data structure.
 * - bool search(word) Returns true if there is any string in the data structure 
 *   that matches word or false otherwise. word may contain dots '.' where dots 
 *   can be matched with any letter.
 * 
 * Time Complexity: 
 *   - addWord: O(m) where m is word length
 *   - search: O(m) for words without '.', O(26^m) worst case with '.'
 * Space Complexity: O(n * m) where n is number of words
 */

#include <iostream>
#include <string>
#include <vector>
using namespace std;

class TrieNode {
public:
    TrieNode* children[26];
    bool isEndOfWord;
    
    TrieNode() {
        isEndOfWord = false;
        for (int i = 0; i < 26; i++) {
            children[i] = nullptr;
        }
    }
};

class WordDictionary {
private:
    TrieNode* root;
    
    bool searchHelper(const string& word, int index, TrieNode* node) {
        if (index == word.length()) {
            return node->isEndOfWord;
        }
        
        char c = word[index];
        
        if (c == '.') {
            // Wildcard: try all possible children
            for (int i = 0; i < 26; i++) {
                if (node->children[i] && searchHelper(word, index + 1, node->children[i])) {
                    return true;
                }
            }
            return false;
        } else {
            // Regular character
            int childIndex = c - 'a';
            if (!node->children[childIndex]) {
                return false;
            }
            return searchHelper(word, index + 1, node->children[childIndex]);
        }
    }
    
public:
    WordDictionary() {
        root = new TrieNode();
    }
    
    void addWord(string word) {
        TrieNode* node = root;
        for (char c : word) {
            int index = c - 'a';
            if (!node->children[index]) {
                node->children[index] = new TrieNode();
            }
            node = node->children[index];
        }
        node->isEndOfWord = true;
    }
    
    bool search(string word) {
        return searchHelper(word, 0, root);
    }
};

int main() {
    WordDictionary wordDict;
    
    wordDict.addWord("bad");
    wordDict.addWord("dad");
    wordDict.addWord("mad");
    
    cout << "search('pad'): " << (wordDict.search("pad") ? "true" : "false") << endl;  // false
    cout << "search('bad'): " << (wordDict.search("bad") ? "true" : "false") << endl;  // true
    cout << "search('.ad'): " << (wordDict.search(".ad") ? "true" : "false") << endl;  // true
    cout << "search('b..'): " << (wordDict.search("b..") ? "true" : "false") << endl;  // true
    cout << "search('b.d'): " << (wordDict.search("b.d") ? "true" : "false") << endl;  // true
    cout << "search('...'): " << (wordDict.search("...") ? "true" : "false") << endl;  // true
    cout << "search('..'): " << (wordDict.search("..") ? "true" : "false") << endl;    // false
    
    return 0;
}

